perception/map_based_prediction/src/map_based_prediction_node.cpp

```diff
@@ -1123,6 +1218,64 @@ void MapBasedPredictionNode::objectsCallback(const TrackedObjects::ConstSharedPt
     "debug/processing_time_ms", processing_time_ms);
 }
 
+void MapBasedPredictionNode::updateCrosswalkUserHistory(
+  const std_msgs::msg::Header & header, const TrackedObject & object, const std::string & object_id)
+{
+  CrosswalkUserData crosswalk_user_data;
+  crosswalk_user_data.header = header;
+  crosswalk_user_data.tracked_object = object;
+  if (crosswalk_users_history_.count(object_id) == 0) {
+    crosswalk_users_history_.emplace(object_id, std::deque<CrosswalkUserData>{crosswalk_user_data});
+  } else {
+    crosswalk_users_history_.at(object_id).push_back(crosswalk_user_data);
+  }
+}
+
+std::string MapBasedPredictionNode::tryMatchNewObjectToDisappeared(
+  const std::string & object_id, std::unordered_map<std::string, TrackedObject> & current_users)
+{
+  if (known_matches_.count(object_id)) {
+    std::string match_id = known_matches_[object_id];
+    // object in the history is already matched to something (possibly itself)
+    if (crosswalk_users_history_.count(match_id)) {
+      // avoid matching two appeared users to one user in history
+      current_users[match_id] = crosswalk_users_history_[match_id].back().tracked_object;
+      return match_id;
+    } else {
+      RCLCPP_WARN_STREAM(
+        get_logger(), "Crosswalk user was " << object_id << "was matched to " << match_id
+                                            << " but the history has deleted. Rematching");
```

Comment:
something like this is cleaner:
```suggestion
const auto match_id_opt = [&]()->std::optional<std::string>{
  if (!known_matches_.count(object_id)) return std::nullopt;
     
    std::string match_id = known_matches_[object_id];
    // object in the history is already matched to something (possibly itself)
    if (crosswalk_users_history_.count(match_id) == 0) {
     RCLCPP_WARN_STREAM(
        get_logger(), "Crosswalk user was " << object_id << "was matched to " << match_id 
        return std::nullopt;}
      // avoid matching two appeared users to one user in history
      current_users[match_id] = crosswalk_users_history_[match_id].back().tracked_object;
      return match_id;
    }
if (  match_id_opt.has_value()) {return  match_id.value()}  
```

Reply:
do you mean use RAII to avoid nested conditionals? I think something like `known_match_ptr` would be a good name here! thank you

Reply:
Well, its an std::optional, not a pointer, so _opt is the correct suffix. It does behave very similarly to a ptr, but the optional is used to specify that you may or may not get a value out of your function. 

The main goal is to avoid nested conditionals as much as possible as you mentioned.

Reply:
Oh sorry didnt mean ptr there, but the comment makes sense . I’ll update the code